File: TheaterSuggestions/Java/SeatsSuggestions/src/main/java/com/baasie/SeatsSuggestions/AuditoriumSeating.java
package com.baasie.SeatsSuggestions;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import lombok.EqualsAndHashCode;

import java.util.Map;

@EqualsAndHashCode
public class AuditoriumSeating {

    private ImmutableMap<String, Row> rows;

    public AuditoriumSeating(Map<String, Row> rows) {
        this.rows = ImmutableMap.copyOf(rows);
    }

    public SeatingOptionSuggested suggestSeatingOptionFor(int partyRequested, PricingCategory pricingCategory) {
        for (Row row : rows.values()) {
            SeatingOptionSuggested seatingOptionSuggested = row.suggestSeatingOption(partyRequested, pricingCategory);

            if (seatingOptionSuggested.matchExpectation()) {
                // Cool, we mark the seat as Allocated (that we turns into a SuggestionMode)
                return seatingOptionSuggested;
            }
        }

        return new SeatingOptionNotAvailable(partyRequested, pricingCategory);
    }

    public AuditoriumSeating allocate(SeatingOptionSuggested seatingOptionSuggested) {
        // Update the seat references in the Auditorium
        return allocateSeats(seatingOptionSuggested.seats());
    }

    private AuditoriumSeating allocateSeats(Iterable<Seat> updatedSeats) {
        Map<String, Row> newVersionOfRows = Maps.newHashMap(rows);

        updatedSeats.forEach(updatedSeat -> {
            Row formerRow = newVersionOfRows.get(updatedSeat.rowName());
            Row newVersionOfRow = formerRow.allocate(updatedSeat);
            newVersionOfRows.put(updatedSeat.rowName(), newVersionOfRow);
        });

        return new AuditoriumSeating(newVersionOfRows);
    }

    public ImmutableMap<String, Row> rows() {
        return rows;
    }
}

File: TheaterSuggestions/CSharp/SeatsSuggestions.Domain/AuditoriumSeating.cs
using System.Collections.Generic;
using Value;
using Value.Shared;

namespace SeatsSuggestions.Domain
{
    public class AuditoriumSeating : ValueType<AuditoriumSeating>
    {
        private readonly Dictionary<string, Row> _rows;
        public IReadOnlyDictionary<string, Row> Rows => _rows;
        public ShowId ShowId { get; }

        public AuditoriumSeating(ShowId showId, Dictionary<string, Row> rows)
        {
            ShowId = showId;
            _rows = rows;
        }

        public SeatingOptionSuggested SuggestSeatingOptionFor(SuggestionRequest suggestionRequest)
        {
            foreach (var row in _rows.Values)
            {
                var seatingOption = row.SuggestSeatingOption(suggestionRequest);

                if (seatingOption.MatchExpectation()) return seatingOption;
            }

            return new SeatingOptionNotAvailable(suggestionRequest);
        }

        public AuditoriumSeating Allocate(SeatingOptionSuggested seatingOptionSuggested)
        {
            // Update the seat references in the Auditorium
            var newAuditorium = AllocateSeats(seatingOptionSuggested.Seats);

            return newAuditorium;
        }

        protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
        {
            return new object[] {new DictionaryByValue<string, Row>(_rows)};
        }

        private AuditoriumSeating AllocateSeats(IEnumerable<Seat> updatedSeats)
        {
            var newVersionOfRows = new Dictionary<string, Row>(_rows);

            foreach (var updatedSeat in updatedSeats)
            {
                var formerRow = newVersionOfRows[updatedSeat.RowName];
                var newVersionOfRow = formerRow.Allocate(updatedSeat);
                newVersionOfRows[updatedSeat.RowName] = newVersionOfRow;
            }

            return new AuditoriumSeating(ShowId, newVersionOfRows);
        }
    }
}